### PR TITLE
Added drawArc to Canvas.

### PR DESCRIPTION
**Pull request by dragostis**
Added a way to call SkCanvas::drawArc from Dart. This method is
optimized for the arc case and should be faster that drawing paths.

### CODE REVIEW

File: lib/ui/painting/canvas.cc [hunk 1]
```diff
@@ -242,6 +243,24 @@ void Canvas::drawCircle(double x,
...
+             double bottom,
+             double startAngle,
+             double sweepAngle,
+             bool useCenter,
+             const Paint& paint,
+             const PaintData& paint_data) {
+  if (!canvas_)
+    return;
+  canvas_->drawArc(SkRect::MakeLTRB(left, top, right, bottom),
+                   startAngle * 180.0 / M_PI,
```
Comment (chinmaygarde) [inline lib/ui/painting/canvas.cc]: I don't think this layer is meant to have or enforce opinions about argument types. Should just forward the arguments as is to Skia. We can perform these conversions at a higher level.

Comment (Hixie) [inline lib/ui/painting/canvas.cc]: The arguments in the Dart API should be in radians.

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -1297,6 +1297,29 @@ class Canvas extends NativeFieldWrapperClass2 {
...
                    ByteData paintData) native "Canvas_drawCircle";
 
+  /// Draw an arc scaled to fit inside the given rectangle. It starts from
+  /// startAngle radians around the oval up to startAngle + sweepAngle
+  /// radians around the oval, with zero radians being the point on
+  /// the right hand side of the oval that crosses the horizontal line
+  /// that intersects the center of the rectangle and with positive
+  /// angles going clockwise around the oval. If useCenter is true, the oval
+  /// center is inserted into the implied path before the arc and the path is
+  /// closed back to the center, forming a wedge. Otherwise, the implied path
```
Comment (Hixie) [inline lib/ui/painting.dart]: "the oval center is inserted into the implied path before the arc and the path is closed back to the center" is an implementation detail, instead describe the output without reference to paths.

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -1297,6 +1297,29 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  /// contains just the arc and is not closed.
```
Comment (Hixie) [inline lib/ui/painting.dart]: same with the next sentence

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -1297,6 +1297,29 @@ class Canvas extends NativeFieldWrapperClass2 {
...
+  void drawArc(Rect rect, double startAngle, double sweepAngle, bool useCenter, Paint paint) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: what happens if useCenter is false and paint says to fill? Does it close the path and fill?

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -1297,6 +1297,29 @@ class Canvas extends NativeFieldWrapperClass2 {
                    List<dynamic> paintObjects,
                    ByteData paintData) native "Canvas_drawCircle";
 
+  /// Draw an arc scaled to fit inside the given rectangle. It starts from
+  /// startAngle radians around the oval up to startAngle + sweepAngle
```
Comment (Hixie) [inline lib/ui/painting.dart]: docs need to define if angles go clockwise or counter-clockwise.

File: lib/ui/painting.dart (hunk 4 above)
Comment (dragostis) [inline lib/ui/painting.dart]: It should have the same behavior as `Path`.

File: lib/ui/painting.dart (hunk 5 above)
Comment (dragostis) [inline lib/ui/painting.dart]: It currently says "with positive angles going clockwise around the oval".

File: lib/ui/painting/canvas.cc (hunk 1 above)
Comment (dragostis) [inline lib/ui/painting/canvas.cc]: @chinmaygarde Based my assumption on https://github.com/flutter/engine/blob/master/lib/ui/painting/path.cc#L136.

File: lib/ui/painting.dart [hunk 6]
```diff
@@ -1302,10 +1302,9 @@ class Canvas extends NativeFieldWrapperClass2 {
   /// radians around the oval, with zero radians being the point on
   /// the right hand side of the oval that crosses the horizontal line
   /// that intersects the center of the rectangle and with positive
-  /// angles going clockwise around the oval. If useCenter is true, the oval
-  /// center is inserted into the implied path before the arc and the path is
+  /// angles going clockwise around the oval. If useCenter is true, the path is
   /// closed back to the center, forming a wedge. Otherwise, the implied path
```
Comment (Hixie) [inline lib/ui/painting.dart]: this still mentions the "implied path" :-)

**Comment by dragostis** [comment]
Some traces: [Archive.zip](https://github.com/flutter/engine/files/464956/Archive.zip)

**Comment by dragostis** [comment]
Related to #3599 and https://bugs.chromium.org/p/skia/issues/detail?id=5227.

**Comment by chinmaygarde** [comment]
Though we updated the Skia revision quite recently, just want to make absolutely sure that you checked with https://chromium.googlesource.com/skia/+/4f3a0ca85d28a8fc7fcc1ac5c4a1864c66bdea14 already pulled in.

**Comment by dragostis** [comment]
@chinmaygarde Very good point! Seems like I've been using an older checkout of Skia when I did the traces.

This fix seems to do its job: 
[dart-timeline-2016-8-12.json.zip](https://github.com/flutter/engine/files/468274/dart-timeline-2016-8-12.json.zip) Re-opening.

File: lib/ui/painting.dart [hunk 7]
```diff
@@ -1297,6 +1297,28 @@ class Canvas extends NativeFieldWrapperClass2 {
...
                    ByteData paintData) native "Canvas_drawCircle";
 
+  /// Draw an arc scaled to fit inside the given rectangle. It starts from
+  /// startAngle radians around the oval up to startAngle + sweepAngle
+  /// radians around the oval, with zero radians being the point on
+  /// the right hand side of the oval that crosses the horizontal line
+  /// that intersects the center of the rectangle and with positive
+  /// angles going clockwise around the oval. If useCenter is true, the arc is
+  /// closed back to the center, forming a circle sector. Otherwise, the arc is
+  /// not closed, forming a circle segment.
```
Comment (Hixie) [inline lib/ui/painting.dart]: This should probably mention that it is faster than creating a path to draw the arc, if that's true.

**Comment by Hixie** [comment]
LGTM with the above comment fix

**Pull request merged**